### PR TITLE
Add Retrieve and Rank

### PR DESCRIPTION
**Pull request by sccheng**
### Summary

This pull request adds the Retrieve and Rank service, addressing issue #328.

### CODE REVIEW

File: Source/RetrieveAndRankV1/Models/RankerDetails.swift [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    /// The unique identifier for this ranker.
+    public let rankerID: String
+    
+    /// The link to this ranker.
+    public let url: String
+    
+    /// The user-supplied name for this ranker.
+    public let name: String
+    
+    /// The date and time, in Coordinated Universal Time, that the ranker was created.
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/RankerDetails.swift]: Updated coordinated universal time to "UTC"

File: Source/RetrieveAndRankV1/Models/Ranking.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    /// The unique identifier of the answer in the collection.
+    public let answerID: String
+    
+    /// The rank of this answer, compared to other answers. A higher value represents a higher
+    /// relevance. The highest score is the sum of the number of potential answers.
+    public let score: Int
+    
+    /// A decimal percentage from 0 to 1, describing the preference for this answer. A higher value
+    /// represents a higher confidence.
+    public let confidence: Int
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/Ranking.swift]: Type should be double

File: Source/RetrieveAndRankV1/Models/Ranking.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+/** An answer and its associated ranking score. */
+public struct RankedAnswer: JSONDecodable {
+    
+    /// The unique identifier of the answer in the collection.
+    public let answerID: String
+    
+    /// The rank of this answer, compared to other answers. A higher value represents a higher
+    /// relevance. The highest score is the sum of the number of potential answers.
+    public let score: Int
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/Ranking.swift]: Make sure this is actually an int

File: Source/RetrieveAndRankV1/Models/Ranking.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    
+    /// A decimal percentage from 0 to 1, describing the preference for this answer. A higher value
+    /// represents a higher confidence.
+    public let confidence: Int
+    
+    /// Used internally to initialize an `RankedAnswer` model from JSON.
+    public init(json: JSON) throws {
+        answerID = try json.string("answer_id")
+        score = try json.int("score")
+        confidence = try json.int("confidence")
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/Ranking.swift]: Update these to json.double

File: Source/RetrieveAndRankV1/Models/SearchAndRankResponse.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+import Foundation
+import Freddy
+
+/** The response received when searching a specific query within the Solr cluster and collection,
+ returned in ranked order. */
+public struct SearchAndRankResponse: JSONDecodable {
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/SearchAndRankResponse.swift]: Add the responseHeader

File: Source/RetrieveAndRankV1/Models/SearchAndRankResponse.swift [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    
+    /// The number of results found.
+    public let numFound: Int
+    
+    /// The index the given results start from.
+    public let start: Int
+    
+    /// The highest ranking score out of the potential answers.
+    public let maxScore: Int
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/SearchAndRankResponse.swift]: Make sure this isn't a double

File: Source/RetrieveAndRankV1/Models/SolrCluster.swift [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+public struct SolrCluster: JSONDecodable {
+    
+    /// The unique identifier for this cluster.
+    public let solrClusterID: String
+    
+    /// The name that identifies this cluster.
+    public let solrClusterName: String
+    
+    /// The size of the cluster. Ranges from 1 to 7.
+    public let solrClusterSize: String
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Models/SolrCluster.swift]: Update the size to be an Int?

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 8]
```diff
@@ -0,0 +1,924 @@
...
+/**
+ The IBM Watson Retrieve and Rank service combines two information retrieval components into a 
+ single service. The service uses Apache Solr in conjunction with a machine learning algorithm to
+ provide users with more relevant search results by automatically re-ranking them.
+ */
+public class RetrieveAndRank {
+    private let username: String
+    private let password: String
+    private let serviceURL: String
+    private let userAgent = buildUserAgent("watson-apis-ios-sdk/0.3.1 RetrieveAndRankV1")
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Update to the correct version

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 9]
```diff
@@ -0,0 +1,924 @@
...
+     
+     - parameter name: The name for the new Solr cluster.
+     - parameter size: The size of the Solr cluster to create. This can range from 1 to 7. You can 
+            create one small free cluster for testing by keeping this value empty.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed with a `SolrCluster` object.
+     */
+    public func createSolrCluster(
+        name: String,
+        size: String? = nil,
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Update the size parameter to an Int

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 10]
```diff
@@ -0,0 +1,924 @@
...
+                }
+            }
+    }
+    
+    /**
+     Gets all configurations for the specific cluster.
+     
+     - parameter solrClusterID: The ID of the cluster that you want the configurations of.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed with an array of `SolrConfig` objects.
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Update this to match the actual return

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 11]
```diff
@@ -0,0 +1,924 @@
...
+    /**
+     Uploads a configuration .zip file set with the given name to the specified cluster.
+     
+     - parameter solrClusterID: The ID of the cluster whose configuration you want to update.
+     - parameter configName: The name of the configuration you want to update.
+     - parameter zipFile: The zip file configuration set that you would like to upload.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed if no error occurs.
+     */
+    public func createSolrConfiguration(
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: update name to uploadConfiguration()

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 12]
```diff
@@ -0,0 +1,924 @@
...
+                
+                let temporaryURL = NSURL(string: "")!
+                let fileURL = destination(temporaryURL, response)
+                success(fileURL)
+            }
+
+    }
+    
+    /**
+     Uploads a configuration .zip file set with the given name to the specified cluster.
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: add a note: configuration writer type must be defined as "json"

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 13]
```diff
@@ -0,0 +1,924 @@
...
+     - parameter writerType: Specifies the QueryResponseWriter to use to process the request. Valid 
+            values are declared in the solrconfig.xml file.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed if no error occurs.
+     */
+    public func createSolrCollection(
+        solrClusterID: String,
+        name: String,
+        configName: String,
+        writerType: String,
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: remove the writerType parameter

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 14]
```diff
@@ -0,0 +1,924 @@
...
+        failure: (NSError -> Void)? = nil,
+        success: (Void -> Void)? = nil) {
+        
+        // construct query parameters
+        var queryParameters = [NSURLQueryItem]()
+        queryParameters.append(NSURLQueryItem(name: "action", value: "CREATE"))
+        queryParameters.append(NSURLQueryItem(name: "wt", value: writerType))
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: set the "wt" to "json"

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 15]
```diff
@@ -0,0 +1,924 @@
...
+     - parameter name: The name of the collection.
+     - parameter writerType: Specifies the QueryResponseWriter to use to process the request. Valid
+            values are declared in the solrconfig.xml file.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed if no error occurs.
+     */
+    public func deleteSolrCollection(
+        solrClusterID: String,
+        name: String,
+        writerType: String,
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: remove the writerType

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 16]
```diff
@@ -0,0 +1,924 @@
...
+     
+     - parameter solrClusterID: The ID of the cluster whose collections you want.
+     - parameter writerType: Specifies the QueryResponseWriter to use to process the request. Valid
+            values are declared in the solrconfig.xml file.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed with an array of collection names.
+     */
+    public func getSolrCollections(
+        solrClusterID: String,
+        writerType: String,
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: remove writerType

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 17]
```diff
@@ -0,0 +1,924 @@
...
+            values are declared in the solrconfig.xml file.
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed with a `SearchResponse` object.
+     */
+    public func search(
+        solrClusterID: String,
+        collectionName: String,
+        query: String,
+        returnFields: String,
+        writerType: String,
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: remove writerType

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 18]
```diff
@@ -0,0 +1,924 @@
...
+     - parameter failure: A function executed if an error occurs.
+     - parameter success: A function executed with a `SearchAndRankResponse` object.
+     */
+    public func searchAndRank(
+        solrClusterID: String,
+        collectionName: String,
+        rankerID: String,
+        query: String,
+        returnFields: String,
+        writerType: String,
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: remove writerType

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 19]
```diff
@@ -0,0 +1,924 @@
...
+                case .Success(let collections): success(collections)
+                case .Failure(let error): failure?(error)
+                }
+            }
+    }
+    
+    /**
+     Update a collection by adding content to it. This indexes the documents and allows us to 
+     search the newly uploaded data later. For more information about the accepted file types and
+     howto structure the content files, refer to this link:
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Update the comment

File: Source/RetrieveAndRankV1/Tests/RetrieveAndRankTests.swift [hunk 20]
```diff
@@ -0,0 +1,924 @@
...
+        retrieveAndRank.deleteSolrConfiguration(
+            "invalid_cluster_ID",
+            configName: "someConfiguration",
+            failure: failure,
+            success: failWithResult)
+        
+        waitForExpectations()
+    }
+    
+//    /** Get a Solr configuration that does not exist. */
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Tests/RetrieveAndRankTests.swift]: remove this test

File: Source/RetrieveAndRankV1/Tests/RetrieveAndRankTests.swift [hunk 21]
```diff
@@ -0,0 +1,924 @@
...
+        waitForExpectations()
+    }
+    
+//    /** Get all configurations when passing an inaccessible Solr cluster ID. */
+//    func testGetConfigurationsWithInaccessibleSolrClusterID() {
+//        let description = "Get all configurations when passing an inaccessible Solr cluster ID."
+//        let expectation = expectationWithDescription(description)
+//        
+//        let failure = { (error: NSError) in
+//            XCTAssertEqual(error.code, 403)
```
Comment (sccheng) [inline Source/RetrieveAndRankV1/Tests/RetrieveAndRankTests.swift]: Change error code to match Alamofire's error.

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 22]
```diff
@@ -452,22 +482,18 @@ public class RetrieveAndRank {
...
         solrClusterID: String,
         name: String,
-        writerType: String,
         failure: (NSError -> Void)? = nil,
         success: (Void -> Void)? = nil) {
         
         // construct query parameters
         var queryParameters = [NSURLQueryItem]()
         queryParameters.append(NSURLQueryItem(name: "action", value: "DELETE"))
-        queryParameters.append(NSURLQueryItem(name: "wt", value: writerType))
```
Comment (glennrfisher) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Should this be `queryParameters.append(NSURLQueryItem(name: "wt", value: "json"))`?

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 23]
```diff
@@ -350,23 +374,33 @@ public class RetrieveAndRank {
...
             }
 
     }
     
     /**
-     Uploads a configuration .zip file set with the given name to the specified cluster.
+     Uploads a configuration .zip file set with the given name to the specified cluster. 
+     
+     Note: in order for your service instance to work with this SDK, you must make sure to define 
+     the writer type in your solrconfig.xml file to be "json".
```
Comment (glennrfisher) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Great comment.

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift [hunk 24]
```diff
@@ -827,11 +854,25 @@ public class RetrieveAndRank {
...
+            return
+        }
+        
         // execute REST request
         Alamofire.upload(
             request,
             multipartFormData: { multipartFormData in
-                multipartFormData.appendBodyPart(fileURL: resultsFile, name: "resultsFile")
+                multipartFormData.appendBodyPart(fileURL: resultsFile, name: "answer_data")
+                multipartFormData.appendBodyPart(data: answerMetadata, name: "answer_metadata")
```
Comment (glennrfisher) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Hmm... this looks good to me. Did you say that it always returns 10 answers regardless of the `numberOfResults` argument?

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift (hunk 22 above)
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Ahh good catch, not sure why it built properly haha.

File: Source/RetrieveAndRankV1/RetrieveAndRank.swift (hunk 24 above)
Comment (sccheng) [inline Source/RetrieveAndRankV1/RetrieveAndRank.swift]: Yup that's correct. For some reason, it almost seems to ignore the `answer_metadata` field; regardless of what number I pass in, I'll always get 10 answers in response.

**Comment by glennrfisher** [comment]
Great updates, @sccheng! This looks great. :)

**Pull request merged**